File: balancing-parentheses-master/solutions/balancing_parentheses.js
function balancingParentheses(string) {
  let missing = 0;
  let openings = 0;

  for (let i = 0; i < string.length; ++i) {
    if (string[i] === '(') {
      ++openings;
      continue;
    }

    if (openings > 0) {
      --openings;
    } else {
      ++missing;
    }
  }

  return missing + openings;
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 0");
  console.log(balancingParentheses('(()())'));

  console.log("");

  console.log("Expecting: 2");
  console.log(balancingParentheses('()))'));

  console.log("");

  console.log("Expecting: 1");
  console.log(balancingParentheses(')'));

  console.log("");

  console.log("Expecting: 0");
  console.log(balancingParentheses('()'));

  console.log("");

  console.log("Expecting: 1");
  console.log(balancingParentheses('('));

  console.log("");

  console.log("Expecting: 2");
  console.log(balancingParentheses(')('));

  console.log("");

  console.log("Expecting: 1");
  console.log(balancingParentheses(')()'));

  console.log("");

  console.log("Expecting: 2");
  console.log(balancingParentheses(')((((()))((())))()(()))('));

  console.log("");

  console.log("Expecting: 3");
  console.log(balancingParentheses(')))'));

  console.log("");

  console.log("Expecting: 3");
  console.log(balancingParentheses('((('));
}

module.exports = balancingParentheses;

function balancingParentheses(parentheses) {
  let count = 0;

  for (let i = 0; i < parentheses.length; i++) {
    if (parentheses[i] === '(') {
      count++;
    } else if (parentheses[i] === ')') {
      count--;
    }
  }

  return Math.abs(count);
}

console.log("Expecting: 3");
console.log(balancingParentheses(')))'));



// Please add your pseudocode to this file
/*/ The `balancingParentheses` function counts the number of
 * unbalanced parentheses in a given string and returns the
 * absolute value of the count.
 * @param parentheses - The `parentheses` parameter is a string
 * that represents a sequence of parentheses. In this case, the
 * string is `')))'`.
 * @returns The function `balancingParentheses` is returning the
 * absolute value of the count variable, which represents the
 * difference between the number of opening and closing
 * parentheses in the input string.
 */
/************************************************************************************** 
* initialize missing to 0 (will store unmatched closing parens count)
* initialize openings to 0 (will store opening parens count)
*
* iterate over string:
*   if char == '(':
*     increment openings
*   else:
*     if openings is 0:
*       increment missing
*     else:
*       decrement openings
*
* return missing + openings


// And a written explanation of your solution
// We can calculate the number of parentheses needed by counting the number of opening
// parentheses that occur in the string and decrementing that count any time a closing
//parenthesis is encountered after that. If we encounter a closing parenthesis and there
//are no opening parentheses (openings = 0), we add to missing. Once we've iterated 
//over the whole string, we just need to add the missing count with the openings count,
// since the openings count will track any opening parentheses for which there were no
*/

File: balancing-parentheses-master/solutions/balancing_parentheses.rb
# The `balancing_parentheses` function counts the number of unbalanced parentheses in a given
# string.
# 
# Args:
#   string: The parameter "string" is a string that contains parentheses. The function
#   "balancing_parentheses" counts the number of unbalanced parentheses in the string.
#
# Returns:
#   The number of unbalanced parentheses in the string.

def balancing_parentheses(string)
  count = 0
  stack = []

  string.each_char do |char|
    if char == '('
      stack.push(char)
    elsif char == ')'
      if stack.empty?
        count += 1
      else
        stack.pop
      end
    end
  end

  count
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 0"
  puts balancing_parentheses('(()())')

  puts

  puts "Expecting: 2"
  puts balancing_parentheses('()))')

  puts

  puts "Expecting: 1"
  puts balancing_parentheses(')')

  # Don't forget to add your own!
end
# Please add your pseudocode to this file
# And a written explanation of your solution

  def balancing_parentheses(string)
  count = 0
  stack = []

  string.each_char do |char|
    if char == '('
      stack.push(char)
    elsif char == ')'
      if stack.empty?
        count += 1
      else
        stack.pop
      end
    end
  end

  count
  end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 0"
  puts balancing_parentheses('(()())')

  puts

  puts "Expecting: 2"
  puts balancing_parentheses('()))')

  puts

  puts "Expecting: 1"
  puts balancing_parentheses(')')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
####
   # The `balancing_parentheses` function counts the number of unbalanced parentheses in a given
   # string.
   # 
   # Args:
   #   string: The parameter "string" is a string that contains parentheses. The function
   # "balancing_parentheses" counts the number of unbalanced parentheses in the string.
# And a written explanation of your solution
# The code is defining a method called `balancing_parentheses` that takes a string as an argument.
 # Inside the method, it initializes a variable `count` to "0" and creates an empty array called
 # `stack`. It then iterates over each character in the string using the `each_char` method.

  ##
  # The `balancing_parentheses` function counts the number of unbalanced parentheses in a given
  # string.
  # 
  # Args:
  #   string: The parameter "string" is a string that contains parentheses.
  